include/grpc/grpc.h

```diff
@@ -387,7 +386,8 @@ grpc_call *grpc_channel_create_registered_call(
    completion of type 'tag' to the completion queue bound to the call.
    The order of ops specified in the batch has no significance.
    Only one operation of each type can be active at once in any given
-   batch. */
+   batch. 
+   THREAD SAFETY: grpc_call_start_batch is thread-compatible */
```

Comment:
So what I said is more lenient that what you said.

What C# is doing is fine. A possible refactoring of what you're doing would
be to have two locks, one for reading, one for writing - and this would
still work - reads are thread-compatible, and writes are thread-compatible.
If you were starting a batch that did both, you'd need to take both mutexes
of course.

On Mon, Jun 22, 2015 at 10:57 AM Jan Tattermusch <notifications@github.com>
wrote:

> In include/grpc/grpc.h
> <https://github.com/grpc/grpc/pull/2047#discussion_r32963275>:
>
> > @@ -387,7 +386,8 @@ grpc_call *grpc_channel_create_registered_call(
> >     completion of type 'tag' to the completion queue bound to the call.
> >     The order of ops specified in the batch has no significance.
> >     Only one operation of each type can be active at once in any given
> > -   batch. */
> > +   batch.
> > +   THREAD SAFETY: grpc_call_start_batch is thread-compatible */
>
> I'd prefer to know that what C# is doing will be supported in the future.
> ;-)
> Could you rephrase the restriction so I understand what you've actually
> meant? I can't guarantee at all from which thread the op will be call.
> Because the original docs said the call methods thread compatible I only
> initiate batched from within a lock owned by a call object, which is my
> understanding of how "thread-compatible" api method should be used.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grpc/grpc/pull/2047/files#r32963275>.
>